rpc/src/client/event_listener.rs

```diff
@@ -83,33 +88,84 @@ impl EventListener {
         Ok(())
     }
 
-    /// Get the next event from the websocket
-    pub async fn get_event(&mut self) -> Result<Option<ResultEvent>, RPCError> {
-        let msg = self
-            .socket
-            .next()
-            .await
-            .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
+    /// Get the next event from the websocket. Automatically handles websocket
+    /// protocol details, like responding to ping messages, so it can either
+    /// produce events or errors.
+    pub async fn get_event(&mut self) -> Result<ResultEvent, RPCError> {
+        loop {
+            let msg = self
+                .socket
+                .next()
+                .await
+                .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
 
-        if let Ok(result_event) = serde_json::from_str::<WrappedResultEvent>(&msg.to_string()) {
-            // if we get an rpc error here, we will bubble it up:
-            return Ok(Some(result_event.into_result()?));
+            match msg {
+                Message::Text(msg_text) => {
+                    match serde_json::from_str::<WrappedResultEvent>(msg_text.as_str()) {
+                        // if we get an rpc error here, we will bubble it up:
+                        Ok(result_event) => return Ok(result_event.into_result()?),
+                        Err(e) => {
+                            return Err(RPCError::new(
+                                Code::Other(-1),
+                                Some(format!(
+                                    "failed to decode incoming message as an event: {}",
+                                    e
+                                )),
+                            ))
+                        }
+                    }
+                }
+                Message::Ping(_) => {
+                    self.socket
+                        .send(Message::Pong(Vec::new()))
+                        .await
+                        .or_else(|e| {
+                            Err(RPCError::websocket_error(format!(
+                                "failed to send pong response: {}",
+                                e
+                            )))
+                        })?
+                }
+                Message::Pong(_) => (),
+                Message::Close(_) => return Err(RPCError::websocket_error("web socket closed")),
+                Message::Binary(_) => {
```

Comment:
Are we expecting binary messages from the WebSocket endpoint?

Reply:
Not that I'm aware of. Everything should be JSON (text) or these Ping msgs.

Reply:
Seems like we shouldn't be expecting any binary messages: https://github.com/tendermint/tendermint/blob/master/rpc/jsonrpc/server/ws_handler.go#L395